### ISSUE TITLE
[BUG] ValueError: assignment destination is read-only

### ISSUE THREAD

**Issue by jesuinovieira** [autogluon/autogluon#3875]
**Bug report checklist**

- [x] I provided code that demonstrates a minimal reproducible example.  
- [x] I confirmed bug exists on the latest mainline of AutoGluon via source install.  
- [x] I confirmed bug exists on the latest stable version of AutoGluon.  

**Describe the bug**

When attempting to consecutively train multiple models, the code encounters an issue after the first model.

**Expected behavior**

The code should successfully train both models.

**To Reproduce**

<details>

```python
import autogluon.tabular
import sklearn.datasets
import tempfile
from sklearn.model_selection import train_test_split


def train(task_type: str) -> None:
    # Load data
    dataset = "diabetes" if task_type == "reg" else "iris"
    bunch = getattr(sklearn.datasets, f"load_{dataset}")(as_frame=True)
    train, tuning = train_test_split(bunch.frame, test_size=0.3, random_state=42)

    print(f"Training model for '{dataset}' ({task_type})")

    # Initialize model
    model = autogluon.tabular.TabularPredictor(
        problem_type="multiclass" if task_type == "clf" else "regression",
        label="target",
        path=tempfile.TemporaryDirectory(dir="/tmp").name,
        verbosity=1
    )

    # Train model
    model.fit(
        train_data=train,
        tuning_data=tuning,
        # https://github.com/autogluon/autogluon/issues/1126
        use_bag_holdout=True,
        time_limit=60,
        presets=["high_quality", "optimize_for_deployment"],
        excluded_model_types=["KNN"],
    )


if __name__ == "__main__":
    train("reg")
    train("clf")
```

</details>

**Screenshots / Logs**

<details>

```txt
Training model for 'diabetes' (reg)
WARNING: Attempted to disable saving of bagged fold models when `use_bag_holdout=True`. Forcing `save_bag_folds=True` to avoid errors.
Deleting model LightGBMXT_BAG_L1. All files under /tmp/tmpc4mtstsh/models/LightGBMXT_BAG_L1 will be removed.
Deleting model LightGBM_BAG_L1. All files under /tmp/tmpc4mtstsh/models/LightGBM_BAG_L1 will be removed.
Deleting model RandomForestMSE_BAG_L1. All files under /tmp/tmpc4mtstsh/models/RandomForestMSE_BAG_L1 will be removed.
Deleting model CatBoost_BAG_L1. All files under /tmp/tmpc4mtstsh/models/CatBoost_BAG_L1 will be removed.
Deleting model ExtraTreesMSE_BAG_L1. All files under /tmp/tmpc4mtstsh/models/ExtraTreesMSE_BAG_L1 will be removed.
Deleting model NeuralNetFastAI_BAG_L1. All files under /tmp/tmpc4mtstsh/models/NeuralNetFastAI_BAG_L1 will be removed.
Deleting model XGBoost_BAG_L1. All files under /tmp/tmpc4mtstsh/models/XGBoost_BAG_L1 will be removed.
Deleting model NeuralNetTorch_BAG_L1. All files under /tmp/tmpc4mtstsh/models/NeuralNetTorch_BAG_L1 will be removed.
Deleting model WeightedEnsemble_L2. All files under /tmp/tmpc4mtstsh/models/WeightedEnsemble_L2 will be removed.

Training model for 'iris' (clf)
WARNING: Attempted to disable saving of bagged fold models when `use_bag_holdout=True`. Forcing `save_bag_folds=True` to avoid errors.
Traceback (most recent call last):
  File "/home/ubuntu/dev/./reproduce-issue.py", line 43, in <module>
    train("clf")
  File "/home/ubuntu/dev/./reproduce-issue.py", line 30, in train
    model.fit(
  File "/home/ubuntu/dev/.venv/lib/python3.10/site-packages/autogluon/core/utils/decorators.py", line 31, in _call
    return f(*gargs, **gkwargs)
  File "/home/ubuntu/dev/.venv/lib/python3.10/site-packages/autogluon/tabular/predictor/predictor.py", line 1099, in fit
    num_stack_levels, time_limit = self._dynamic_stacking(**ds_args, ag_fit_kwargs=ag_fit_kwargs, ag_post_fit_kwargs=ag_post_fit_kwargs)
  File "/home/ubuntu/dev/.venv/lib/python3.10/site-packages/autogluon/tabular/predictor/predictor.py", line 1186, in _dynamic_stacking
    stacked_overfitting = self._sub_fit_memory_save_wrapper(
  File "/home/ubuntu/dev/.venv/lib/python3.10/site-packages/autogluon/tabular/predictor/predictor.py", line 1329, in _sub_fit_memory_save_wrapper
    stacked_overfitting = _ds_ray.get(finished[0])
  File "/home/ubuntu/dev/.venv/lib/python3.10/site-packages/ray/_private/auto_init_hook.py", line 24, in auto_init_wrapper
    return fn(*args, **kwargs)
  File "/home/ubuntu/dev/.venv/lib/python3.10/site-packages/ray/_private/client_mode_hook.py", line 103, in wrapper
    return func(*args, **kwargs)
  File "/home/ubuntu/dev/.venv/lib/python3.10/site-packages/ray/_private/worker.py", line 2524, in get
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(ValueError): ray::_sub_fit() (pid=898190, ip=88.99.184.74)
  File "/home/ubuntu/dev/.venv/lib/python3.10/site-packages/autogluon/tabular/predictor/predictor.py", line 4842, in _sub_fit
    predictor._fit(ag_fit_kwargs=ag_fit_kwargs, ag_post_fit_kwargs=ag_post_fit_kwargs)
  File "/home/ubuntu/dev/.venv/lib/python3.10/site-packages/autogluon/tabular/predictor/predictor.py", line 1115, in _fit
    self._learner.fit(**ag_fit_kwargs)
  File "/home/ubuntu/dev/.venv/lib/python3.10/site-packages/autogluon/tabular/learner/abstract_learner.py", line 159, in fit
    return self._fit(X=X, X_val=X_val, **kwargs)
  File "/home/ubuntu/dev/.venv/lib/python3.10/site-packages/autogluon/tabular/learner/default_learner.py", line 93, in _fit
    X, y, X_val, y_val, X_unlabeled, holdout_frac, num_bag_folds, groups = self.general_data_processing(X, X_val, X_unlabeled, holdout_frac, num_bag_folds)
  File "/home/ubuntu/dev/.venv/lib/python3.10/site-packages/autogluon/tabular/learner/default_learner.py", line 198, in general_data_processing
    X_val[self.label].replace([np.inf, -np.inf], np.nan, inplace=True)
  File "/home/ubuntu/dev/.venv/lib/python3.10/site-packages/pandas/core/generic.py", line 7919, in replace
    return self._update_inplace(result)
  File "/home/ubuntu/dev/.venv/lib/python3.10/site-packages/pandas/core/generic.py", line 4822, in _update_inplace
    self._maybe_update_cacher(verify_is_copy=verify_is_copy, inplace=True)
  File "/home/ubuntu/dev/.venv/lib/python3.10/site-packages/pandas/core/series.py", line 1418, in _maybe_update_cacher
    ref._maybe_cache_changed(cacher[0], self, inplace=inplace)
  File "/home/ubuntu/dev/.venv/lib/python3.10/site-packages/pandas/core/frame.py", line 4439, in _maybe_cache_changed
    self._mgr.iset(loc, arraylike, inplace=inplace)
  File "/home/ubuntu/dev/.venv/lib/python3.10/site-packages/pandas/core/internals/managers.py", line 1085, in iset
    return self._iset_single(
  File "/home/ubuntu/dev/.venv/lib/python3.10/site-packages/pandas/core/internals/managers.py", line 1270, in _iset_single
    blk.set_inplace(slice(iloc, iloc + 1), value, copy=copy)
  File "/home/ubuntu/dev/.venv/lib/python3.10/site-packages/pandas/core/internals/blocks.py", line 1039, in set_inplace
    self.values[locs] = values
ValueError: assignment destination is read-only
```

</details>

**Installed Versions**

<details>

```txt
INSTALLED VERSIONS
------------------
date                   : 2024-01-22
time                   : 19:57:39.075128
python                 : 3.10.12.final.0
OS                     : Linux
OS-release             : 5.15.0-91-generic
Version                : #101-Ubuntu SMP Tue Nov 14 13:30:08 UTC 2023
machine                : x86_64
processor              : x86_64
num_cores              : 8
cpu_ram_mb             : 31336.5390625
cuda version           : None
num_gpus               : 0
gpu_ram_mb             : []
avail_disk_size_mb     : 139288

accelerate             : 0.21.0
async-timeout          : 3.0.1
autogluon              : 1.0.1b20240101
autogluon.common       : 1.0.1b20240101
autogluon.core         : 1.0.1b20240101
autogluon.features     : 1.0.1b20240101
autogluon.multimodal   : 1.0.1b20240101
autogluon.tabular      : 1.0.1b20240101
autogluon.timeseries   : 1.0.1b20240101
boto3                  : 1.34.23
catboost               : 1.2.2
defusedxml             : 0.7.1
evaluate               : 0.4.1
fastai                 : 2.7.13
gluonts                : 0.14.3
hyperopt               : 0.2.7
imodels                : None
jinja2                 : 3.1.3
joblib                 : 1.3.2
jsonschema             : 4.17.3
lightgbm               : 4.1.0
lightning              : 2.0.9.post0
matplotlib             : 3.8.2
mlforecast             : 0.10.0
networkx               : 3.2.1
nlpaug                 : 1.1.11
nltk                   : 3.8.1
nptyping               : 2.4.1
numpy                  : 1.26.3
nvidia-ml-py3          : 7.352.0
omegaconf              : 2.2.3
onnxruntime-gpu        : None
openmim                : 0.3.9
orjson                 : 3.9.12
pandas                 : 2.1.4
Pillow                 : 10.2.0
psutil                 : 5.9.8
PyMuPDF                : None
pytesseract            : 0.3.10
pytorch-lightning      : 2.0.9.post0
pytorch-metric-learning: 1.7.3
ray                    : 2.6.3
requests               : 2.28.2
scikit-image           : 0.20.0
scikit-learn           : 1.4.0
scikit-learn-intelex   : None
scipy                  : 1.12.0
seqeval                : 1.2.2
setuptools             : 60.2.0
skl2onnx               : None
statsforecast          : 1.4.0
statsmodels            : 0.14.1
tabpfn                 : None
tensorboard            : 2.15.1
text-unidecode         : 1.3
timm                   : 0.9.12
torch                  : 2.0.1+cpu
torchmetrics           : 1.1.2
torchvision            : 0.15.2+cpu
tqdm                   : 4.65.2
transformers           : 4.31.0
utilsforecast          : 0.0.10
vowpalwabbit           : None
xgboost                : 2.0.3
```

</details>

**Comment by yanhong-zhao-ef** [comment]
any updates on this?

**Comment by Innixma** [comment]
I ran the provided code snippet without issue on latest mainline.

Can you confirm if the issue is resolved on your end if using AutoGluon 1.1 or latest mainline?

**Comment by yanhong-zhao-ef** [comment]
I was passing in pandas dataframe to the estimator so I ended up passing in ```ds_args={"memory_safe_fits": False}``` to disable spawning multiple ray processes with auto-stacking on (using preset best quality) so the bug was gone for me on 1.0.0. Will remove that to try if see 1.1.0 solves the problem out of box but this line X_val[self.label].replace([np.inf, -np.inf], np.nan, inplace=True) is tricky when you try to edit in place across different processes

**Comment by yanhong-zhao-ef** [comment]
can report back the issue is solved with 1.1 and the training is lot of faster!

**Comment by jesuinovieira** [comment]
Confirmed, the issue is resolved on the latest release (1.1.0)
